libsupport/src/EntityTypeManager.cpp

```diff
@@ -476,54 +476,24 @@ katana::EntityTypeManager::PrintEntityTypes() const {
 bool
 katana::EntityTypeManager::Equals(
     const katana::EntityTypeManager& other) const {
```

Comment:
If we are weakening the semantics of this to not include subtyping relations (which is the only part you are missing I think, since differences in numbering are not semantically meaningful), then I think we should rename this function. Maybe: `TestEquivalent` or something.

I think we should reserve `Equals` for real equality. Since people don't like using `operator==` for that.

Reply:
One thing that reinforces my thought is that I'm introducing tools to automatically wrap "conventional" methods on classes. I think `Equals` should be one of them.

Reply:
+1 to this

Reply:
Can you explain the subtyping point?
I think `City:Place` is a subtype of `City`, but is that not necessarily so?

Reply:
My point was just that the subtyping relationships need to be checked for anything that could be considered true equality. You are only checking the set of names if I read correctly.

Reply:
+1 to `*Equivalent`

I didn't understand the subtyping point as well. This checks entity-type-IDs are the same (including the atomic types). We could relax it further to say that entity-types are the same but the IDs for those could be different but I don't think that's required.

Reply:
I renamed this function to TypeEquivalent and I added this comment.
Note, I don't understand what this comment means.
`// This relation is TypeEquivalent, not Equals because it does not test subtype relationships.`

Reply:
This had some unanticipated downstream, e.g., Equals being used in SanityCheckPartition.cpp and g-serialize.cpp, and v3-uint16-entity-type-ids.cpp.  I'm changing them to TypeEquivalent, but I'm not sure if the previous version of Equals was better suited for their use.  Of course, I can have both functions, but I was seeing spurious problems with the previous definition when I recreate an existing graph from the empty graph.

Reply:
> I renamed this function to TypeEquivalent and I added this comment.
> Note, I don't understand what this comment means.
> `// This relation is TypeEquivalent, not Equals because it does not test subtype relationships.`

OK. Let me try again:

An `EntityTypeManager` conceptually stores two kinds of information:

1. A set of types (concretely, names and IDs), and
2. The "inheritance heirarchy", or more formally the "subtype relation". This is a set of super types for each non-atomic (intersection) type. This is equivalent to having lists of subtypes, though that is less obvious from the implementation.

The function seem to check that the set of type names are the same, but does not check that the super type lists are the same. Though maybe I totally misunderstood.

Reply:
We do NOT store anything in the class for (2). It is derived from (1); that is (1) implies (2). So I don't see what check this function is missing. Can you please see if there is a (counter-)example?

Reply:
You definitely store The supertyping data: `entity_type_id_to_atomic_entity_type_ids_`.

As for your second point, you are right. You do actually check subtyping, but via strings. Sorry. My bad. My comment about subtyping was incorrect.

Sorry.

Reply:
Thanks for the clarification!

Reply:
Thanks for the conversation. I assume then that y'all are ok with me leaving this method titled Equals.

Reply:
I'm on the fence. The problem is that the API does expose the IDs and this method explicitly *doesn't* check if IDs are the same. So if the client code only uses type names, then yes this is equality, but if the client code uses IDs, it's not. If there is any real way to get IDs from outside this class (PropertyGraph for instance) then the user could easily be using IDs directly making this equality invalid. For context: in OO the normal concept of equality is sustainability. So things are equal if you could substitute one for the other.

I think I would prefer this be called "Equivalent" (or even better "Isomorphic" ;-) ). That would also avoid my python wrappers automatically exposing this as `==` in Python which is probably not a great idea for the reasons above.

Reply:
Fair enough. I can get behind that sentiment.